extensions/oplogPopulator/allocationStrategy/ImmutableConnector.js

```diff
@@ -0,0 +1,22 @@
+const AllocationStrategy = require('./AllocationStrategy');
+
+/**
+ * @class ImmutableConnector
+ *
+ * @classdesc ImmutableConnector is an allocation
+ * strategy that always requires a new connector to be
+ * created for each bucket.
+ */
+class ImmutableConnector extends AllocationStrategy {
+    /**
+     * Get best connector for assigning a bucket
+     * @param {Connector[]} connectors available connectors
+     * @returns {Connector} connector
+     */
+    // eslint-disable-next-line no-unused-vars
+    getConnector(connectors) {
+        return null;
```

Comment:
code-wise, it seems like a good place to return the "existing" connector (if any) which still had the extraneous connector

Reply:
Isn't this already done by the allocator? We keep track of the buckets->connectors, and here we try to rely on the real list (not the "in memory" list), so we already do that

Reply:
i mean the `Allocator` keeps track of already "allocated" buckets (not the really applied once) - synchronously ; but it could be the strategy's role to remember that the bucket (even though not allocated anymore) is still "pending" in some connector.

`getConnector()` is called when we need to find the connector to use for a 'new' bucket (i.e. one we don't already know/track), so it would be fitting to return the existing connector if any.

→ instead of modifying the Allocator, it would be very nice if we could just use a "wrapper" Strategy, which returns the existing connector is the bucket was "pending" deallocation; and just passes the call to underlying connector otherwise.

Reply:
If I understand correctly, we want to avoid the new logic in the Allocator, and always remove the bucket from the map (i.e., we do not consider buckets "in use" in the Allocator), and we then rely on the strategy to automatically get the current connector, the one where the bucket is still pending, or a new connector?

While it would simplify the code, I do not see the difference in terms of complexity with the comment above, as we would need to parse all the connectors to detect if one has the bucket in it?

Reply:
we don't want to "avoid the new logic in the Allocator" ; we want to leverage the interface to abstract implementation details, so that we add additional, composable logic in separate class instead of making a bigger and bigger class which covers too many aspects to be easily understandable.

Each class should end up with their own responsability (somewhat as they already have), and clear contract between them.

so there may be some logic added in Allocator (typically the "create connector if getConnector returns null" logic is what would be expected: the contract has been extended to support a bit more) ; and maybe the contract needs to be extended to cover the whole semantic